### ISSUE TITLE
Exposing the control values.

### ISSUE THREAD

**Issue by andersio** [ReactiveCocoa/ReactiveCocoa#3229]
The Property contract has an assumed one-to-one relationship between a value and a change event so far, inferred from our documentations and test cases.

> The current value of a property can be obtained from the value getter. The producer getter returns a signal producer that will send the property’s current value, followed by **all** changes over time.

On the other hand, `UIControl` and `NSControl` does not have such one-to-one relationship. Changes are emitted as control events or notifications only if they are initiated by UI interactions.

Even if we manage to reconstruct the relationship by exotic means (which is possible for just AppKit as far as I've tried), it may cause infinite feedback loops in cases that alter the value on-the-fly in response to user interactions.

However, it is no doubt that they fit into the Property interface. So should we live with the discrepancy, and say that Property implementation may choose _not to emit all the changes_?

Related PR: https://github.com/ReactiveCocoa/ReactiveSwift/pull/52

/cc @ReactiveCocoa/reactiveswift

**Comment by liscio** [comment]
This is a big concern I hit up against when I first started working with these APIs, and later hit it harder when I started the Bindings work initially. The inputs and outputs of UI controls are not always symmetric (in their types as well as behavior), and UI controls should not store data.

We had a discussion earlier on Slack about "inlets" and "outlets" for controls, where "inlets" are represented by `BindingTarget` and "outlets" would be best represented as `Signal`s. Frankly, by forcing a control to expose anything resembling a `Property` means that the API users should (reasonably) expect the control to store values. ***I feel strongly that this is the wrong approach.***

Instead, if we limited the bindings to controls to be more restrictive in that they expose discrete inlets and outlets the we can force the user of controls (including the framework itself, eventually) to build up a set of machinery to deal with these discrete behaviors.

Consider the example of a text field in an inspector of a drawing program. The inspector follows a user's selection of circles on a canvas. The text field displays the radius of the currently-selected circle.

When the user changes their selected circle, the value can flow into an "object controller" and then propagate up into the text field. The user enters a new value that flows into the "object controller", (optionally) gets validated, and propagates back down into the underlying model. If validation fails, the "object controller" pushes the last-known good value back up into the control.

Now consider if we exposed a property on the text field. It's very likely that an API user would commit one of the following bugs:

* Hook the property up to the model directly, which could *potentially* lead to the value of the control getting applied to their next selected object if the API user is not careful.

* "Setter loops" where the property value gets set on the control, and bounces back down into the model. (Maybe, with some controls, but this requires us to create "special properties" that don't actually behave like properties should behave. E.g. https://github.com/ReactiveCocoa/ReactiveSwift/pull/52)

In the latter case, why try and force the control to expose a property at all? 

In conclusion, I don't feel that UI controls are the appropriate place to convey the storage of values using a `Property` of any kind. I feel that the right thing to do is supply a set of discrete `Signal` paths for controls that allow user input, as well as the associated `BindingTarget` that we've exposed.

**Comment by andersio** [comment]
`ControlSubject`* is a proposed solution, as a result of the discussion with @liscio and @sharplet on Slack. It acts as _both_ a binding target and a signal, but it does not conform to the property protocol.

https://github.com/ReactiveCocoa/ReactiveCocoa/pull/3210/commits/dc3a6db71b0dcf6f69a207ed885f1c9c10a78c9a

* As usual, better names are welcomed.

**Comment by mdiep** [comment]
I'd suggest that we take 1 concrete example to work through its API. #3210 is pretty unwieldy, which makes it hard to absorb. It also requires an actual implementation.

What control would be a good example to work through? `UIButton`? Let's post the existing Rex API here and then discuss what it should be.

**Comment by liscio** [comment]
Just for posterity's sake, an alternate was also proposed, where controls would expose a set of `inlet`s and `outlet`s with the underlying `Signal`s and `BindingTarget`s. For instance:

```
// Assuming an NSTextField, NSButton
userNameTextField.inlet.stringValue <~ viewModel.userName
loginButton.inlet.isEnabled <~ viewModel.userNameValid

viewModel.userName <~ userNameTextField.outlet.stringValue
viewModel.userNameValid <~ userNameTextField.outlet.continuousStringValue.map { $0.characters.count > 4 }
```

I preferred that `inlet` (or `input`) and `outlet` would get separated for self-documenting API, and to group "what goes in" and "what comes out". But others argued (rightly, I think) that a "beefier" endpoint type would allow us to expose a two-way binding in the future.

**Comment by andersio** [comment]
```
extension UIButton {
    public var rex_pressed: MutableProperty<CocoaAction> { get }
    public var rex_title: MutableProperty<String> { get }
}
```

```
extension UISwitch {
    public var rex_on: MutableProperty<Bool> { get }
}
```

**Comment by mdiep** [comment]
At a minimum, I think we want something that lets you:

1. Receive notifications when the value changes due to user input
2. Let's you bind something to it to change the value

That sounds an awful lot like a property to me. To conform to `PropertyProtocol`, we'd only need to provide a way to get the current value—which can easily be done in a way that gets the current value from the control. I think it's fine to have a caveat that you only get notified when the user changes the value—not from programmatic changes.

If we want it to be something other than a property, that's okay with me. But as I said elsewhere, I think it's pragmatic to conform to `PropertyProtocol`.

**Comment by liscio** [comment]
> That sounds an awful lot like a property to me. To conform to PropertyProtocol, we'd only need to provide a way to get the current value—which can easily be done in a way that gets the current value from the control.

This is precisely what I'm hoping to avoid.

As I understand things, `MutableProperty` encapsulates the concept of both storage & notifications, sharing the `SignalProducer` semantics. That is, there is a concept of "the current value" as well as "future values."

Furthermore, there is an implied storage of some sort. Whether an underlying "storage object" is being wrapped, or it's a concrete `MutableProperty`, there should be an expectation that the thing acting as a property is a unit of storage.

With controls, there should _never_ be a current value that enters the discussion. We need to think of them as a combination of "fire & forget side effects" that are displayed to the user, and a stream of "user-inputted values" that never tell us the last one the user entered. That's up to our ViewModel to store (in a `MutableProperty`!) if we choose.

Here's another scenario I just dreamed up. Imagine that in IB you set some placeholder text on a NSTextField such as, `"Hello, world!"`. `stringValue` is a `MutablePropertyProtocol` as you proposed. In your `viewDidLoad`:

```
viewModel.userName <~ textField.stringValue
textField.stringValue <~ viewModel.userName
```

In the above scenario, your model gets updated with the "Hello, world!" placeholder text from IB because we decided to grab the current value from the UI widget in our design.

```
textField.stringValue <~ viewModel.userName
viewModel.userName <~ textField.stringValue
```

Flipping the order of operations "fixes the problem", but **there shouldn't be a problem to begin with!**

If we implement the above as either a discrete type that encapsulates a `Signal`,`BindingTarget` pair, or a pair of discrete `Signal` and `BindingTarget` endpoints, then either of the above orderings will be correct.

**Comment by mdiep** [comment]
Okay, that makes sense to me. Thanks for explaining. 👍 

There was a lot of discussion about this in Slack today. What's the current proposal?

**Comment by liscio** [comment]
@mdiep I think that we're currently hovering around the idea of keeping things very simple, where we use the names of exposed extensions to distinguish signals from binding targets. For instance, on `NSTextField`:

``` swift
// These supply changes originating from the user
var stringValues: Signal<String, NoError> { get }
var continuousStringValues: Signal<String, NoError> { get }

// These act as binding targets, and present incoming values to the user
var string: BindingTarget<String> { get }
var isEnabled: BindingTarget<Bool> { get }
```

@andersio suggested that we could handle awkward plurals by tacking on `Signal`, but I managed to convince him that`Values` might work better from a readability standpoint. For instance:

``` swift
// "Update the user name with the text field's text signal"
viewModel.userName <~ textField.reactive.textSignal

// "Update the user name with the text field's text values"
viewModel.userName <~ textField.reactive.textValues
```

That's where we seem to have landed at the moment, and so far it appears to be working in the limited amount I've been using it.

**Comment by andersio** [comment]
@liscio May I make a slight correction: the nomenclature in the PR follows the names in the base type. So for UIKit text controls, it is `texts` and `continuousTexts`. Likewise, we have `stringValues`, `continuousStringValues`, `integerValues`, etc, in `NSControl`.

`Values` is only appended when the key is not a countable noun, e.g. `isOnValues`.

**Comment by mdiep** [comment]
I don't think _text_ is countable in this context, so the `Value` suffix would apply.

This all sounds good to me. 👍

**Comment by liscio** [comment]
Yeah I'm with @mdiep. The plural of "text" doesn't work here. Super happy to hear we're all mostly on the same page! 👍 🌟

**Comment by mdiep** [comment]
![English is weird!](https://media.giphy.com/media/TtzVUVNBNlXLG/giphy.gif)

**Comment by sharplet** [comment]
I think `reactive.` might be enough to supply the context of a stream of values in many cases. So instead of `isOnValues`, which reads awkwardly to me, I think `reactive.isOn` could be sufficient.

Likewise for `text`, I feel like `reactive.text` and `reactive.continuousText` could be good names.

What do you think?

**Comment by mdiep** [comment]
The issue is that we have two items: a `BindingTarget` and a `Signal`. They can't both have the `text` name.

**Comment by sharplet** [comment]
Right, and that was the reason behind having a type that conforms to both protocols.

**Comment by andersio** [comment]
Just want to note that if [`ControlSubject`](https://github.com/ReactiveCocoa/ReactiveCocoa/pull/3210#pullrequestreview-2895682) is to be revived in some other names, we need to also convert all existing method signatures with `Signal` to be constrained by `SignalProtocol` for the new type to work everywhere.

**Comment by andersio** [comment]
Here is my two-way binding prototype: https://gist.github.com/andersio/4988f9ee44d65f27cdbd0076a89d6a21

As shown in the code snippets, a single type can pack a known combination of targets and signals for bindings with a neat `<~>` operator.
